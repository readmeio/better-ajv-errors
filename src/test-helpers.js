import { readFileSync } from 'fs';
import path from 'path';

export async function getSchemaAndData(name, dirPath) {
  const schemaPath = path.join(dirPath, '..', '__fixtures__', name, 'schema.json');
  const schema = JSON.parse(readFileSync(schemaPath, 'utf8'));
  const dataPath = path.join(dirPath, '..', '__fixtures__', name, 'data.json');
  const json = readFileSync(dataPath, 'utf8');
  const data = JSON.parse(json);

  return [schema, data, json];
}
