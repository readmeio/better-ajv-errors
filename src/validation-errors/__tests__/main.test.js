import Ajv from 'ajv/dist/2020';
import betterAjvErrors from '../..';
import { getSchemaAndData } from '../../test-helpers';
import { openapi } from '@apidevtools/openapi-schemas';

describe('Main', () => {
  it('should support js output format for default errors', async () => {
    const [schema, data] = await getSchemaAndData('default', __dirname);

    const ajv = new Ajv();
    const validate = ajv.compile(schema);
    const valid = validate(data);
    expect(valid).toBe(false);

    const res = betterAjvErrors(schema, data, validate.errors, {
      format: 'js',
    });
    expect(res).toMatchSnapshot();
  });

  it('should support js output format for required errors', async () => {
    const [schema, data] = await getSchemaAndData('required', __dirname);

    const ajv = new Ajv();
    const validate = ajv.compile(schema);
    const valid = validate(data);
    expect(valid).toBe(false);

    const res = betterAjvErrors(schema, data, validate.errors, {
      format: 'js',
    });
    expect(res).toMatchSnapshot();
  });

  it('should support js output format for additionalProperties errors', async () => {
    const [schema, data] = await getSchemaAndData('additionalProperties', __dirname);

    const ajv = new Ajv();
    const validate = ajv.compile(schema);
    const valid = validate(data);
    expect(valid).toBe(false);

    const res = betterAjvErrors(schema, data, validate.errors, {
      format: 'js',
    });
    expect(res).toMatchSnapshot();
  });

  // Though this is testing `patternProperties` cases, the error that'll come back will just be for `pattern` since
  // that's what `patternPropeties` utilizes in the schema.
  it('should support js output format for patternProperties errors', async () => {
    const [, data] = await getSchemaAndData('patternProperties', __dirname);

    // The OpenAPI 3.1 schema has a good example of this for component names so we're using that instead of writing
    // our own.
    const schema = openapi.v31;
    const ajv = new Ajv({ allErrors: true, strict: false, validateFormats: false });
    const validate = ajv.compile(schema);
    const valid = validate(data);
    expect(valid).toBe(false);

    const res = betterAjvErrors(schema, data, validate.errors, {
      format: 'js',
    });
    expect(res).toMatchSnapshot();
  });

  it('should support js output format for unevaluatedProperties errors', async () => {
    const [schema, data] = await getSchemaAndData('unevaluatedProperties', __dirname);

    const ajv = new Ajv();
    const validate = ajv.compile(schema);
    const valid = validate(data);
    expect(valid).toBe(false);

    const res = betterAjvErrors(schema, data, validate.errors, {
      format: 'js',
    });
    expect(res).toMatchSnapshot();
  });

  it('should support js output format for enum errors', async () => {
    const [schema, data] = await getSchemaAndData('enum', __dirname);

    const ajv = new Ajv();
    const validate = ajv.compile(schema);
    const valid = validate(data);
    expect(valid).toBe(false);

    const res = betterAjvErrors(schema, data, validate.errors, {
      format: 'js',
    });
    expect(res).toMatchSnapshot();
  });
});
