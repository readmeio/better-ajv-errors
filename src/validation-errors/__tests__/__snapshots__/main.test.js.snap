// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should support js output format for additionalProperties errors 1`] = `
Array [
  Object {
    "end": Object {
      "column": 27,
      "line": 1,
      "offset": 26,
    },
    "error": " Property baz is not expected to be here",
    "path": "",
    "start": Object {
      "column": 22,
      "line": 1,
      "offset": 21,
    },
  },
]
`;

exports[`Main should support js output format for default errors 1`] = `
Array [
  Object {
    "end": undefined,
    "error": "/id: type must be number",
    "path": "/id",
    "start": Object {
      "column": 7,
      "line": 1,
      "offset": 6,
    },
  },
]
`;

exports[`Main should support js output format for enum errors 1`] = `
Array [
  Object {
    "end": Object {
      "column": 12,
      "line": 1,
      "offset": 11,
    },
    "error": "/id must be equal to one of the allowed values: foo, bar",
    "path": "/id",
    "start": Object {
      "column": 7,
      "line": 1,
      "offset": 6,
    },
    "suggestion": "Did you mean bar?",
  },
]
`;

exports[`Main should support js output format for patternProperties errors 1`] = `
Array [
  Object {
    "end": Object {
      "column": 261,
      "line": 1,
      "offset": 260,
    },
    "error": "/components/securitySchemes Property \\"Basic Access Authentication\\" must match pattern ^[a-zA-Z0-9._-]+$",
    "path": "/components/securitySchemes",
    "start": Object {
      "column": 244,
      "line": 1,
      "offset": 243,
    },
  },
  Object {
    "end": undefined,
    "error": "/components/securitySchemes: propertyNames property name must be valid",
    "path": "/components/securitySchemes",
    "start": Object {
      "column": 262,
      "line": 1,
      "offset": 261,
    },
  },
]
`;

exports[`Main should support js output format for required errors 1`] = `
Array [
  Object {
    "error": "/nested must have required property 'id'",
    "path": "/nested",
    "start": Object {
      "column": 11,
      "line": 1,
      "offset": 10,
    },
  },
]
`;

exports[`Main should support js output format for unevaluatedProperties errors 1`] = `
Array [
  Object {
    "end": Object {
      "column": 27,
      "line": 1,
      "offset": 26,
    },
    "error": " Property baz is not expected to be here",
    "path": "",
    "start": Object {
      "column": 22,
      "line": 1,
      "offset": 21,
    },
  },
]
`;
